### PR TITLE
Do not update project details after sync jobs

### LINKED ISSUE #2835: "sync" type project updates mark their project as failed
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
 - API

##### SUMMARY
Pre-run project syncs are not supposed to ordinarily make changes to the project they use.

Thus, a failure in a project sync should not cause the project to be marked failed, just as how it does not update its last update link usually.

##### ENVIRONMENT
* AWX version: X.Y.Z
* AWX install method: openshift, minishift, docker on linux, docker for mac, boot2docker
* Ansible version:  X.Y.Z
* Operating System:
* Web Browser:

##### STEPS TO REPRODUCE

Run a JT and sigkill the process while the project sync is running

##### EXPECTED RESULTS

Project not marked as failed

##### ACTUAL RESULTS

project marked as failed

### PR DESCRIPTION
**Pull request by AlanCoding**
##### SUMMARY
Resolves https://github.com/ansible/awx/issues/2835

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
 - API

##### AWX VERSION

```

```

### CODE REVIEW

**Review by matburt — APPROVED** [review]

**Comment by AlanCoding** [comment]
For information, this does have a connection to https://github.com/ansible/awx/pull/2719

Presumably, in the old system, a failed sync could be righted by a subsequent successful sync. But that old scheme seems like it doesn't respect the authority of "check" type project updates, which are supposed to determine the revision, and @wenottingham indicated that a failed state should reset (or make invalid) the revision.

Anyway, this all kind of comes back to the change where we allowed offline updates

I think that change was https://github.com/ansible/awx/pull/567, I really like the behavior, but some permutations related to this are still being tweaked

File: docs/clustering.md [hunk 1]
```diff
@@ -335,6 +335,9 @@ It's important to note that not all instances are required to be provisioned wit
 Project updates behave differently than they did before. Previously they were ordinary jobs that ran on a single instance. It's now important that
 they run successfully on any instance that could potentially run a job. Project's will now sync themselves to the correct version on the instance immediately
 prior to running the job.
+When the sync happens, it is recorded in the database as a project update with a `launch_type` of "sync"
+and a `job_type` of "run". Project syncs will not change the status or version of the project,
+but will, instead, update the source tree only on the instance where they run.
```
Comment (chrismeyersfsu) [inline docs/clustering.md]: * Unless it is the first project update ran. Said differently, unless there are no other previous project updates, of any launch_type or job_type, for the project.

Comment (AlanCoding) [reply docs/clustering.md]: valid to include, I added a blurb about it.

**Review by chrismeyersfsu — APPROVED** [review]

**Comment by Testicule** [comment]
So this clustering is the reason that AWX contacts SCM server every time a job is run, even when "Update on Launch is not selected" for a project?

This is the reason that AWX has been crippled by a requirement that it do a checkout every time a job is launched?

Really?

What is the problem you're trying to solve here?
You have some tower nodes which may be left with an old checkout, if a project update was initiated, but not all the nodes successfully updated? Is that the problem? 

If so, why don't you just fix that problem, instead of creating a raft of additional problems by kludging things so the project is checked out every time a job is launched. That is just nonsense.

You need some logic within your cluster to keep it in sync, or so it knows when the nodes are not in sync. You don't cripple every node and standalone instance by kludging in an unnecessary project update before launching every job, just to mask this bug.

**Comment by chrismeyersfsu** [comment]
We could require NFS to keep the project dirs in sync. Would that solve your problem @Testicule ?

**Comment by Testicule** [comment]
Keeping the project dirs in sync is not my problem. I've got a single instance of AWX which is being crippled by these unnecessary checkouts every time i run a job. That's my problem.

But if you think you need NFS to keep your project dirs in sync that's clearly jumping to another wrong conclusion. You don;t need another mechanism to keep them in sync - all you need is a mechanism to know when they're in sync or out of sync. That could be done via other means. Your nodes should be able to work out amongst themselves if their dirs are in sync without contacting SCM, and if they are, you don't need to contact SCM.

**Comment by chrismeyersfsu** [comment]
If the project points at master, how can you know you have the most up-to-date master without asking the remote?

Also, what about non-distributed SCM types?

**Comment by Testicule** [comment]
What is going on here? Is there some sort of contagious, collective brain damage happening and I'm the only one who hasn't contracted it yet?

We are talking about launching job templates which do *not* have "Update revision on Launch" selected. That means we want to launch them and *WE DO NOT CARE IF WE DON'T HAVE THE LATEST COPY FROM SCM*.

That is what that button means.

You have projects.
They have a button to pull down the latest revision from SCM.
When you press that button, that's what should happen.
When you don't press that button, and the templates you launch do *not* have "Update revision on Launch" selected, the project should *NOT* be checked out, or updated, or anything.

I apologise if my comments seem rude but i am seriously flabbergasted. 
Why can't people grasp this?

When you press the button to update a project, you go and contact SCM and checkout the latest revision. If you have multiple tower nodes in a cluster, if they need to, they can talk amongst themselves and inform each other whether they were successful in pulling down the latest updates from SCM. If a node fails, then that node should know it has an outdated copy and needs to update before it can run jobs. All of this should be implemented in the logic to checkout/update a project, if you want all nodes in a cluster to have the same source code checked out. If you want to complicate things further with distributed SCMs, or worry about the fact that one node may update a little earlier or later than others, and get different revision of the project, if it was changed during the update process, then that is up to you to manage as part of your process to update a project. Have one node check the current revision, notify other nodes of that specific revision to checkout, other nodes then grab that specific revision, any node that fails can't run jobs until it is successful. Or whatever. Make it as complicated and bulletproof as you like. That's up to you.

When you press the button to update a project from SCM, that is when this happens. It does *not* need to happen when a job is launched from a template that has "Update on Launch" *DE*selected.

When i launch a job that  "Update on Launch" *DE*selected. I want that job to launch. That's what "Launch" means. Blast off. Woosh, Off it goes. It doesn't mean "Initiate a cluster-wide project update, re-import inventories, and waste a bunch of time before even attempting to launch the job."

As people have pointed out, if you insist on checking out the project before running every job, you not only waste time, you also make it impossible for the tower to manage the hosts running your SCM, and people, including myself, also want to do that. If my Tower/AWX has the latest revsion, and i know it has the latest revision because i just pressed the button to Update the damned project, then i should be able to launch a job to shutdown my SCM server, take a snapshot, install updates, reboot it. And then i should be able to launch another job to power it on and start it all up. I shouldn't then have that 2nd job fail, because AWX/Tower decides to contact the SCM to checkout my project. I did that already. It is *DONE*. It does not need to be done again. Now i just want to launch the job to restart my SCM server. That is all. No project update. None. Zip. Nada. Diddly.

Is that clear enough?

**Comment by ryanpetrello** [comment]
@Testicule,

AWX is free software.  _Many_ of our conributors are volunteers working on this for free in their spare time.  We expect **all** community members to abide by our community code of conduct.  That means treating **everyone** with the respect they deserve, which you're not doing:

https://docs.ansible.com/ansible/latest/community/code_of_conduct.html

> Every community inevitably has disagreements, but remember that it is possible to disagree respectfully and courteously. Disagreements are never an excuse for rudeness, hostility, threatening behavior, abuse (verbal or physical), or personal attacks.

This is not the type of discourse we tolerate:

> What is going on here? Is there some sort of contagious, collective brain damage happening and I'm the only one who hasn't contracted it yet?

Get it under control or show yourself the door.

**Pull request merged**